### PR TITLE
[AMDGPU] Fix gfx12 waitcnt type for image_msaa_load (#90201)

### PR DESCRIPTION
**Pull request by jayfoad**
image_msaa_load is actually encoded as a VSAMPLE instruction and
requires the appropriate waitcnt variant.

### CODE REVIEW

**Comment by jayfoad** [comment]
Let's not backport this yet since @pendingchaos has pointed out a problem with #90201.

**Comment by jayfoad** [comment]
> Let's not backport this yet since @pendingchaos has pointed out a problem with #90201.

Fixed by #90710 which I have added to this PR.

File: llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp [hunk 1]
```diff
@@ -187,8 +187,12 @@ VmemType getVmemType(const MachineInstr &Inst) {
...
   const AMDGPU::MIMGBaseOpcodeInfo *BaseInfo =
       AMDGPU::getMIMGBaseOpcodeInfo(Info->BaseOpcode);
-  return BaseInfo->BVH ? VMEM_BVH
-                       : BaseInfo->Sampler ? VMEM_SAMPLER : VMEM_NOSAMPLER;
+  // We have to make an additional check for isVSAMPLE here since some
+  // instructions don't have a sampler, but are still classified as sampler
+  // instructions for the purposes of e.g. waitcnt.
+  return BaseInfo->BVH                                         ? VMEM_BVH
+         : (BaseInfo->Sampler || SIInstrInfo::isVSAMPLE(Inst)) ? VMEM_SAMPLER
+                                                               : VMEM_NOSAMPLER;
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:195]: Use a temporary variable? This nested ternary formatting is breaking my brain

**Comment by jayfoad** [comment]
Too late to backport - no more 18.x releases are planned.

**Pull request closed without merging**